Implemented GetMedium, GetUnsignedMedium, ReadMedium, ReadUnsignedMedium, WriteMedium, SetMedium

ScarletKuro opened the pull request:
Implemented functions

- https://netty.io/4.0/api/io/netty/buffer/ByteBuf.html#getMedium-int-
- https://netty.io/4.0/api/io/netty/buffer/ByteBuf.html#getUnsignedMedium-int-
- https://netty.io/4.0/api/io/netty/buffer/ByteBuf.html#readMedium--
- https://netty.io/4.0/api/io/netty/buffer/ByteBuf.html#readUnsignedMedium--
- https://netty.io/4.0/api/io/netty/buffer/ByteBuf.html#writeMedium-int-
- https://netty.io/4.0/api/io/netty/buffer/ByteBuf.html#setMedium-int-int-
- https://netty.io/4.0/api/io/netty/buffer/ByteBufUtil.html#swapMedium-int-

Implemented tests

- GetMediumBoundaryCheck1
- GetMediumBoundaryCheck2
- TestSequentialMediumAccess
- TestSequentialMediumLeAccess
- TestSequentialUnsignedMediumAccess
- TestSequentialUnsignedMediumLeAccess
- TestGetMediumAfterRelease
- TestGetMediumLeAfterRelease
- TestGetUnsignedMediumAfterRelease
- TestGetUnsignedMediumLeAfterRelease
- TestSetMediumAfterRelease
- TestSetMediumLeAfterRelease
- TestReadMediumAfterRelease
- TestReadMediumLeAfterRelease
- TestReadUnsignedMediumAfterRelease
- TestReadUnsignedMediumLeAfterRelease
- TestWriteMediumAfterRelease
- TestWriteMediumLeAfterRelease
- TestRandomMediumAccess
- TestRandomMediumLeAccess
- TestRandomUnsignedMediumAccess
- TestRandomUnsignedMediumLeAccess

Related with [#241](https://github.com/Azure/DotNetty/issues/241)

msftclas commented:
@ScarletKuro,
Thanks for having already signed the Contribution License Agreement. Your agreement was validated by Microsoft. We will now review your pull request.
_Thanks,
Microsoft Pull Request Bot_

ScarletKuro commented:
I have revamped the code a bit.
1)Added new class MediumUtil in DotNetty.Common/Utilities(not sure if it's right place?) with ToMediumInt and ToUnsignedMediumInt functions. Since C# doesn't have builtin int24/uint24 and this class will help to convert int to medium, plus it improves the readability of GetMedium, ReadMedium, SwapMedium etc functions.
2)I also added the WriteUnsignedMedium function.

nayato commented on src/DotNetty.Buffers/UnpooledHeapByteBuffer.cs:
```diff
@@ -217,6 +218,19 @@ public override long GetLong(int index)
+        public override int GetMedium(int index)
+        {
+            this.EnsureAccessible();
+            return this._GetMedium(index).ToMediumInt();
```
the fact that you need to call `.ToMediumInt()` here raises a question as to what _GetMedium returns then. Would it make sense to pull `.ToMediumInt()` into _GetMedium?

ScarletKuro replied:
We need to call .ToMediumInt() because in fact _GetMedium returns Unsigned Int24. And in java code the _GetMedium is _GetUnsignedMedium in reality. I renamed it to _GeMedium because it was was out of the ordinary in the AbstractByteBuffer all abstracted functions are named as _GetShort, _GetInt, _GetLong and etc..

nayato replied:
while it is not in line with other methods it simplifies conversions across the board. Could you pls unify with the Java code? It is fine for other types, but Medium does not map to framework's native type and signed<->unsigned conversion become confusing and error prone so reducing places where they're needed is a win.

---
In reply to: [114223854](https://github.com/Azure/DotNetty/pull/244#discussion_r114223854) [](ancestors = 114223854)

nayato commented on src/DotNetty.Common/Utilities/MediumUtil.cs:
```diff
@@ -0,0 +1,37 @@
+    public static class MediumUtil
+    {
+        //High byte bit-mask used when a 24-bit integer is stored within a 32-bit integer.
+        const int MediumBitMask = -16777216;
```
pls use `unchecked((int)0xFF000000)` instead

ScarletKuro commented:
Now it should be fine, _GetMedium retruns the signed value.

nayato commented on src/DotNetty.Common/Utilities/MediumUtil.cs:
```diff
@@ -0,0 +1,35 @@
+
+        public static int ToUnsignedMediumInt(this int value)
+        {
+            return Convert.ToInt32((uint)value & ~MediumBitMask);
```
why `Convert.ToInt32(x)` and not just `(int)x`?

nayato commented:
looking good now. A minor comment on Convert.ToInt32 and I'll be happy to merge.

ScarletKuro commented:
Yeah, i agree. Simple cast is much more preferable, there is no need in Convert.ToInt32. Changed to ```int(x)```.
I did rebase and squash, and edited the commit message.

nayato commented:
thanks for seeing this through!
